File: js_version/src/TrainingData.js

'use strict'

//
//
// TrainingData

function TrainingData (str_content) {
	this._arr_lines = str_content.trim().split('\n');
};

TrainingData.prototype.isEof = function() {
	return (this._arr_lines.length == 0)
}

TrainingData.prototype._extract_from_data = function (str_pattern, arr_output) {

	arr_output.length = 0; // clear array

	var str_value = this._arr_lines.shift().trim();
	var arr_elements = str_value.split(' ');

	if (arr_elements.length > 0 &&
		arr_elements[0] == str_pattern)
	{
		for (var i = 1; i < arr_elements.length; ++i)
			arr_output.push( parseFloat(arr_elements[i]) );
	}

	return arr_output.length;
}

TrainingData.prototype.getTopology = function(arr_topology) {
	return this._extract_from_data('topology:', arr_topology);
}

TrainingData.prototype.getNextInputs = function(arr_inputVals) {
	return this._extract_from_data('in:', arr_inputVals);
}

TrainingData.prototype.getTargetOutputs = function(arr_targetOutputVals) {
	return this._extract_from_data('out:', arr_targetOutputVals);
}

// TrainingData
//
//

module.exports = TrainingData;

File: projects/training-logic/src/utilities/TrainingData.cpp

#include "./TrainingData.hpp"

#include <sstream>

TrainingData::TrainingData(const std::string& filename)
{
    m_file_trainingData.open(filename.c_str());

    if (m_file_trainingData.fail()) {
        throw std::invalid_argument("file not found");
    }
}

void TrainingData::getTopology(std::vector<unsigned> &arr_topology)
{
    arr_topology.reserve(10); // pre-allocate

    std::string line;
    std::string label;

    std::getline(m_file_trainingData, line);
    std::stringstream ss(line);
    ss >> label;

    if (this->isEof() || label != "topology:")
    {
        abort();
    }

    while (!ss.eof())
    {
        unsigned n;
        ss >> n;
        arr_topology.push_back(n);
    }
}

unsigned TrainingData::getNextInputs(t_vals &arr_inputVals)
{
    arr_inputVals.clear();

    std::string str_line;
    std::getline(m_file_trainingData, str_line);
    std::stringstream sstr(str_line);

    std::string str_label;
    sstr >> str_label;
    if (str_label == "in:")
    {
        double dbl_oneValue;

        while (sstr >> dbl_oneValue)
            arr_inputVals.push_back(dbl_oneValue);
    }

    return arr_inputVals.size();
}

unsigned TrainingData::getTargetOutputs(t_vals &arr_targetOutputVals)
{
    arr_targetOutputVals.clear();

    std::string str_line;
    std::getline(m_file_trainingData, str_line);
    std::stringstream sstr(str_line);

    std::string str_label;
    sstr >> str_label;
    if (str_label == "out:")
    {
        double dbl_oneValue;

        while (sstr >> dbl_oneValue)
            arr_targetOutputVals.push_back(dbl_oneValue);
    }

    return arr_targetOutputVals.size();
}
